### ISSUE TITLE
Change syntax for declaring terminals with a custom lexer

### ISSUE THREAD

**Issue by malleusinferni** [nikomatsakis/lalrpop#132]
Current syntax:

```rust
enum custom::Token {
    "foo" => custom::Token::Foo,
    "bar" => custom::Token::Bar(<custom::Baz>),
}
```

Alternative 1 ("assign"):

```rust
type Terminal = custom::Token;

"foo" = custom::Token::Foo;
"bar" = custom::Token::Bar(<custom::Baz>);
```

Alternative 2 ("match"):

```rust
Terminal = {
    custom::Token::Foo => "foo",
    custom::Token::Bar(<custom::Baz>) => "bar",
};
```

----

For the most part, LALRPOP grammar rules borrow from two familiar Rust conventions: Assignment (`name = value`) and pattern matching (`input => output`). The terminal declaration syntax mixes both conventions and follows neither. A new user perusing an example grammar is liable to glance over the `enum` section; mistake `name => value` for `input => output`; deduce (incorrectly) that this section contains no useful information; assume that they should look elsewhere to understand how a custom lexer is set up; and, finally, either cry for help or simply abandon all hope.

The essence of my proposal is to make the syntax conform to **one** of the conventions described. I've given two examples of how this might be done. Note that neither of these masquerades as an `enum` declaration block. It's my further recommendation that such a device be reserved for the automatic generation of custom lexers, if and when that feature materializes.

**Comment by nikomatsakis** [comment]
Yes, I agree the `enum` syntax is bad. Among other things, the token need not be an enum! Any type that can be matched with a pattern will do. 

Option 3 isn't complete, since it never specifies the "terminal type", but I think I do like the reversed pattern order. Maybe something like this?

```
type Terminal = XXX;

match { // maybe a "match" section?
    custom::token::Bar(<custom::Baz>) => "foo"
}
```

Hmm. The artificial "match" section doesn't seem great.
